lib/PrintAsClang/DeclAndTypePrinter.cpp

```diff
@@ -763,19 +764,145 @@ class DeclAndTypePrinter::Implementation
 
     // Finish the result type.
     multiPart.finish();
+  }
 
+  /// Print C or C++ trailing attributes for a function declaration.
+  void printFunctionClangAttributes(FuncDecl *FD, FunctionType *funcTy) {
     if (funcTy->getResult()->isUninhabited()) {
       os << " SWIFT_NORETURN";
     } else if (!funcTy->getResult()->isVoid() &&
                !FD->getAttrs().hasAttribute<DiscardableResultAttr>()) {
       os << " SWIFT_WARN_UNUSED_RESULT";
     }
+  }
 
-    printAvailability(FD);
+  void printAbstractFunctionAsCFunction(FuncDecl *FD) {
+    printDocumentationComment(FD);
+    Optional<ForeignAsyncConvention> asyncConvention
+      = FD->getForeignAsyncConvention();
+    Optional<ForeignErrorConvention> errorConvention
+      = FD->getForeignErrorConvention();
+    assert(!FD->getGenericSignature() &&
+           "top-level generic functions not supported here");
+    auto funcTy = FD->getInterfaceType()->castTo<FunctionType>();
+    auto resultTy = getForeignResultType(
+        FD, funcTy, asyncConvention, errorConvention);
 
+    assert(FD->getAttrs().hasAttribute<CDeclAttr>() && "not a cdecl function");
+    os << "SWIFT_EXTERN ";
+    printFunctionDeclAsCFunctionDecl(
+        FD, FD->getAttrs().getAttribute<CDeclAttr>()->Name, resultTy);
+    printFunctionClangAttributes(FD, funcTy);
+    printAvailability(FD);
     os << ";\n";
   }
 
+  struct FuncionSwiftABIInformation {
+    FuncionSwiftABIInformation(FuncDecl *FD, Mangle::ASTMangler &mangler) {
+      isCDecl = FD->getAttrs().hasAttribute<CDeclAttr>();
+      if (!isCDecl) {
+        mangledName = mangler.mangleAnyDecl(FD, /*prefix=*/true);
+      } else {
+        cdeclName = FD->getAttrs().getAttribute<CDeclAttr>()->Name;
+      }
+    }
+
+    StringRef getSymbolName() const {
+      return isCDecl ? cdeclName : mangledName;
+    }
+
+    bool useCCallingConvention() const { return isCDecl; }
+
+    bool useMangledSymbolName() const { return !isCDecl; }
+
+  private:
+    bool isCDecl;
+    std::string mangledName;
+    StringRef cdeclName;
+  };
+
+  FuncionSwiftABIInformation
+  printSwiftABIFunctionSignatureAsCxxFunction(FuncDecl *FD) {
```

Comment:
Maybe add a comment above this and `printAbstractFunctionAsCxxFunctionThunk` to explain roughly what they do and why you would use one vs the other.